File: 15-3sum/3sum.ts
function threeSum(nums: number[]): number[][] {
    const ans = [];
    const numbers = nums.sort((a, b) => a - b);

    const twoSum = (newNums : number[],target : number, startIndex : number) : void => {
        let endIndex = newNums.length - 1;
        while(startIndex < endIndex){

            if(newNums[startIndex] + newNums[endIndex] === target){
                ans.push([-target,newNums[startIndex], newNums[endIndex]]);
                startIndex ++;
                while(startIndex < endIndex && newNums[startIndex] === newNums[startIndex - 1]) startIndex ++;
                endIndex --;
                while(startIndex < endIndex && newNums[endIndex] === newNums[endIndex + 1]) endIndex --;
            }
            else if (newNums[startIndex] + newNums[endIndex] < target) startIndex ++;
            else endIndex --;
        }
    }

    for(let i = 0; i < numbers.length; i ++){
        if(i === 0 || numbers[i] != numbers[i - 1]){
            console.log('start index is ' + (i + 1))
            twoSum(numbers,-numbers[i],i + 1);
        }
    }
    return ans;
};

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            int curTarget = 0 - nums[i];
            if (i != 0 && nums[i] == nums[i - 1])
                continue;
            twoSum(nums, curTarget, i, nums.length - 1, ans);
        }
        return ans;
    }

    private void twoSum(int[] nums, int target, int l, int right, List<List<Integer>> ans) {
    int left = l + 1;

    while (left < right) {
        int sum = nums[left] + nums[right];

        if (sum < target) {
            left++;
        } else if (sum > target) {
            right--;
        } else {
            ans.add(Arrays.asList(nums[l], nums[left], nums[right]));

            left++;
            right--;

            while (left < right && nums[left] == nums[left - 1])
                left++;

            while (left < right && nums[right] == nums[right + 1])
                right--;
        }
    }
}

}